internal/oci/ecr/ecr.go

```diff
@@ -5,50 +5,122 @@ import (
 }
 
-func (r *ECR) CredentialFunc(registry string) auth.CredentialFunc {
-	return r.Credential
+type ecrPublicClient struct {
+	client ECRPublicClient
 }
 
-func (r *ECR) Credential(ctx context.Context, hostport string) (auth.Credential, error) {
-	cfg, err := config.LoadDefaultConfig(context.Background())
+func (r *ecrPublicClient) GetAuthorizationToken(ctx context.Context) (string, time.Time, error) {
+	client := r.client
+	if client == nil {
+		cfg, err := config.LoadDefaultConfig(context.Background())
+		if err != nil {
+			return "", time.Time{}, err
+		}
+		client = ecrpublic.NewFromConfig(cfg)
+	}
+	response, err := client.GetAuthorizationToken(ctx, &ecrpublic.GetAuthorizationTokenInput{})
 	if err != nil {
-		return auth.EmptyCredential, err
+		return "", time.Time{}, err
+	}
+	authData := response.AuthorizationData
+	if authData == nil {
+		return "", time.Time{}, ErrNoAWSECRAuthorizationData
+	}
+	if authData.AuthorizationToken == nil {
+		return "", time.Time{}, auth.ErrBasicCredentialNotFound
 	}
-	r.client = ecr.NewFromConfig(cfg)
-	return r.fetchCredential(ctx)
+	return *authData.AuthorizationToken, *authData.ExpiresAt, nil
 }
 
-func (r *ECR) fetchCredential(ctx context.Context) (auth.Credential, error) {
-	response, err := r.client.GetAuthorizationToken(ctx, &ecr.GetAuthorizationTokenInput{})
+type ecrPrivateClient struct {
+	client ECRClient
+}
+
+func (r *ecrPrivateClient) GetAuthorizationToken(ctx context.Context) (string, time.Time, error) {
+	client := r.client
+	if client == nil {
+		cfg, err := config.LoadDefaultConfig(ctx)
+		if err != nil {
+			return "", time.Time{}, err
+		}
+		client = ecr.NewFromConfig(cfg)
+	}
+	response, err := client.GetAuthorizationToken(ctx, &ecr.GetAuthorizationTokenInput{})
 	if err != nil {
-		return auth.EmptyCredential, err
+		return "", time.Time{}, err
 	}
 	if len(response.AuthorizationData) == 0 {
-		return auth.EmptyCredential, ErrNoAWSECRAuthorizationData
+		return "", time.Time{}, ErrNoAWSECRAuthorizationData
 	}
-	token := response.AuthorizationData[0].AuthorizationToken
+	authData := response.AuthorizationData[0]
 
-	if token == nil {
-		return auth.EmptyCredential, auth.ErrBasicCredentialNotFound
+	if authData.AuthorizationToken == nil {
+		return "", time.Time{}, auth.ErrBasicCredentialNotFound
 	}
+	return *authData.AuthorizationToken, *authData.ExpiresAt, nil
+}
+
+func CredentialFunc(registry string) auth.CredentialFunc {
+	r := &svc{}
+	switch {
+	case strings.HasPrefix(registry, "public.ecr.aws"):
+		r.client = &ecrPublicClient{}
+	default:
+		r.client = &ecrPrivateClient{}
+	}
+	return r.Credential
+}
+
+type svc struct {
+	mu               sync.Mutex
+	expiresAt        time.Time
+	cachedCredential auth.Credential
+	client           Client
+}
+
+func (r *svc) Credential(ctx context.Context, hostport string) (auth.Credential, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if time.Until(r.expiresAt) > 0 {
+		return r.cachedCredential, nil
+	}
```

Comment:
As we don't have control over where/how Flipt is deployed I suggest to always use UTC:
```suggestion
	if time.Now().UTC().Before(r.expiresAt) > 0 {
		return r.cachedCredential, nil
	}
```